### DISCUSSION TITLE
How do I implement Anti-Forgery in Blazor ver 8 InteractiveServer?

### DISCUSSION THREAD

**Discussion by DavidThielen** [dotnet/aspnetcore#54572] [Q&A]
> **Update:** [The document for WebApplication.UseAntiforgery()](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.builder.antiforgeryapplicationbuilderextensions.useantiforgery?view=aspnetcore-8.0#microsoft-aspnetcore-builder-antiforgeryapplicationbuilderextensions-useantiforgery(microsoft-aspnetcore-builder-iapplicationbuilder)) is a tautology. This should say where it needs to be called and if it is sufficient or if anything else also needs to be done.

I am converting a Blazor ver 6/7 Server to ver 8 InteractiveServer.

In ver 6/7 I set up anti-forgery with the following in _Host.cshtml

```razor
@inject IAntiforgery Xsrf
// ...
<body>
    @{
        var initialTokenState = new InitialApplicationState
            {
                XsrfToken = Xsrf.GetAndStoreTokens(HttpContext).RequestToken,
                Cookie = HttpContext.Request.Cookies[".AspNetCore.Cookies"],
                RemoteIp = HttpContext.GetRemoteIpAddress(Logger)
            };
    }

```
I moved the content of _Host.cshtml to App.razor and in App.razor, it has no `HttpContext` property. [This documentation discusses submitting a form](https://learn.microsoft.com/en-us/aspnet/core/blazor/call-web-api?view=aspnetcore-8.0#antiforgery-support), but nothing about setting it up for rendering a page. When I try to run and render a page (no form on it), I get:

```output
InvalidOperationException: Endpoint /sitemap (/sitemap) contains anti-forgery metadata, but a middleware was not found that supports anti-forgery.
Configure your application startup by adding app.UseAntiforgery() in the application startup code. If there are calls to app.UseRouting() and app.UseEndpoints(...), the call to app.UseAntiforgery() must go between them. Calls to app.UseAntiforgery() must be placed after calls to app.UseAuthentication() and app.UseAuthorization().
```
Clearly there are different steps I should be taking for anti-forgery in version 8. Where is this documented? And in my case - no endpoints. My app is going from rendermode ver 6/7 server to ver 8 InteractiveServer. (And a Google search for `useantiforgery` just leads to others asking about all this.)

I have found a way to get my app running. However, **I do not know if this is correct** and **I do not know if this stops forgery attacks**. I'm using what another person found works and they did not seem sure if this is correct. But it works and no one else has answered, so I figure better than no answer. 

I replaced:
```cs
app.UseAuthentication();
app.UseAuthorization();

app.MapControllers();
app.MapBlazorHub();
app.MapFallbackToPage("/_Host");

if (builder.Environment.IsProduction())
    app.MapHealthChecks("health").AllowAnonymous();
```
with:
```cs
app.UseAuthentication();
app.UseAuthorization();

app.UseAntiforgery();
app.MapRazorPages();

app.MapRazorComponents<App>()
    .AddInteractiveServerRenderMode();

if (builder.Environment.IsProduction())
    app.MapHealthChecks("health").AllowAnonymous();
```
If this is all that's required - great move on the part of the Blazor team - a lot easier than what was required before.

**Question 1:** So, is the above the correct way to protect against anti-forgery attacks?

**Question 2:** I've tried to find the instructions on how to stop forgery attacks in versions 6/7 - and cannot find it. Aside from the following, is there anything else that was done the old way that I now need to find and delete?
```razor
@inject IAntiforgery Xsrf
var initialTokenState = new InitialApplicationState ...
<component type="typeof(App)" render-mode="Server" param-InitialState="initialTokenState" />
```
thanks - dave

**Comment by reuben-ahmed** [comment]
I'm getting this in my **Blazor 8 server/client/shared** project as well. 

- It *was* working a month ago, didn't touch the codebase at all. 
- Tried to run it locally this morning and I get:

```
"System.InvalidOperationException: Endpoint / (/) contains anti-forgery metadata, but a middleware was not found that supports anti-forgery."
``` 

I tried upgrading to the latest .NET 8 SDK (8.0.203), same error.  For reference, it was "working" last month on 8.0.103, but now failing on that version too.
- Is this a new .NET issue (how did it start to fail without any code change?)
- Is it a browser issue (chrome/edge behaving the same)
- My server has `app.useHttpsRedirection`, but my `https://localhost:port` won't load. Don't know if that's related...by default, my local blazor is trying to load `http://localhost:port`

If I add the following in the server's Program.cs (after `app.UseRouting`):
`app.UseAntiforgery();`

It seems like my server starts to respond, but my client then gets:
```
InvalidOperationException: Cannot provide a value for property 'Http' on type 'Acme.Client.Pages.Index'. There is no registered service of type 'System.Net.Http.HttpClient'.
```

I can't remove `app.MapControllers()` because my server has REST API controllers being used.

I'm using VSCode with C# Dev Kit extension, if that matters.
- I see that extension was released on `3-12-2024`, any chance it's related to that?
- c# extension was updated on `3-12-2024` also
